From c40e3dfe2b9e48d870e79062730701c3923c5d88 Mon Sep 17 00:00:00 2001
From: xufuji456 <839789740@qq.com>
Date: Mon, 22 May 2023 17:55:38 +0800
Subject: [PATCH] avcodec/mediacodecdec: add low-latency decoding

The low-latency decoding mode has available in Android api 30.
It will keep less buffer internally, and output frame as fast as possible.
See https://developer.android.google.cn/reference/android/media/MediaCodec?hl=en#PARAMETER_KEY_LOW_LATENCY

Signed-off-by: xufuji456 <839789740@qq.com>
---
 libavcodec/mediacodecdec.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/libavcodec/mediacodecdec.c b/libavcodec/mediacodecdec.c
index 21464900d1..5fece6ee5b 100644
--- a/libavcodec/mediacodecdec.c
+++ b/libavcodec/mediacodecdec.c
@@ -394,6 +394,10 @@ static av_cold int mediacodec_decode_init(AVCodecContext *avctx)
     ff_AMediaFormat_setString(format, "mime", codec_mime);
     ff_AMediaFormat_setInt32(format, "width", avctx->width);
     ff_AMediaFormat_setInt32(format, "height", avctx->height);
+    // low-latency mode: available in Android api 30
+    if (avctx->flags & AV_CODEC_FLAG_LOW_DELAY) {
+        ff_AMediaFormat_setInt32(format, "low-latency", 1);
+    }
 
     s->ctx = av_mallocz(sizeof(*s->ctx));
     if (!s->ctx) {
-- 
2.32.0 (Apple Git-132)

